File: agent_kickbacks/tracker.py
"""Client-side impression and click tracking (best-effort)."""

import httpx


class Tracker:
    def __init__(self, server_url: str):
        self.server = server_url.rstrip("/")

    def log_impression(self, ad_id: str, wallet: str) -> None:
        """Report a confirmed display to the ad server (the server is the
        authority on what is billable)."""
        try:
            httpx.post(
                f"{self.server}/ad/impression",
                json={"ad_id": ad_id, "user_wallet": wallet},
                timeout=2.0,
            )
        except httpx.HTTPError:
            pass  # best effort

    def log_click(self, ad_id: str, wallet: str) -> None:
        try:
            httpx.post(
                f"{self.server}/ad/click",
                json={"ad_id": ad_id, "user_wallet": wallet},
                timeout=2.0,
            )
        except httpx.HTTPError:
            pass

File: openclaw-plugin/src/lib/tracker.ts
/**
 * Impression + click tracking. Best-effort, fail-open — the server is the
 * authority on what is billable, so a dropped report just means no earnings.
 */

import { Ad, adId } from "./ad-client.js";

const TIMEOUT_MS = 2000;

async function post(url: string, body: unknown): Promise<void> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
  } catch {
    // best effort
  } finally {
    clearTimeout(timer);
  }
}

/** Report a confirmed display. `impression_token` from /ad/request is required. */
export async function trackImpression(ad: Ad, wallet: string, server: string): Promise<void> {
  await post(`${server}/ad/impression`, {
    ad_id: adId(ad),
    user_wallet: wallet,
    token: ad.impression_token ?? "",
  });
}

export async function trackClick(ad: Ad, wallet: string, server: string): Promise<void> {
  await post(`${server}/ad/click`, { ad_id: adId(ad), user_wallet: wallet });
}
